### PR TITLE
feat: Convert Transport to deal with Envelopes

### PR DESCRIPTION
**Pull request by Swatinem**

### CODE REVIEW

File: sentry-core/src/envelope.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    fn to_buf(envelope: Envelope) -> std::io::Result<Vec<u8>> {
+        let mut vec = vec![];
+        envelope.to_writer(&mut vec)?;
+        Ok(vec)
```
Comment (flub) [inline sentry-core/src/envelope.rs:119]: Since writing to `Vec` is infallible you could remove the `Result` here and unwrap right away.  (Additionally you're just unwrapping this function at al callers anyway afaik)

File: sentry-core/src/envelope.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    #[test]
+    fn test_empty() {
+        assert_eq!(to_buf(Envelope::new()).unwrap(), b"{}\n");
+    }
+
+    #[test]
+    fn test_event() {
+        let event_id = Uuid::parse_str("22d00b3f-d1b1-4b5d-8d20-49d138cd8a9c").unwrap();
```
Comment (flub) [inline sentry-core/src/envelope.rs:129]: Does it make sense to add a test with the event_id set to None as well?

File: sentry-core/src/envelope.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+/// A Sentry Envelope.
+///
+/// An Envelope is the data format that Sentry uses for Ingestion. It can contain
+/// multiple Items, some of which are related, such as Events, and Event Attachments.
+/// Other Items, such as Sessions are independant.
+///
+/// See the [documentation on Envelopes](https://develop.sentry.dev/sdk/envelopes/)
+/// for more details.
+#[derive(Default, Debug, PartialEq)]
```
Comment (flub) [inline sentry-core/src/envelope.rs:30]: Event is clonable, should Envelope also be clonable?

File: sentry-core/src/envelope.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
+use std::io::Write;
+
+use crate::protocol::Event;
```
Comment (flub) [inline sentry-core/src/envelope.rs:3]: Should envelope not also live there?

File: sentry-log/src/integration.rs [hunk 5]
```diff
@@ -42,7 +42,10 @@ impl Integration for LogIntegration {
             log::set_max_level(filter);
         }
 
-        INIT.call_once(|| log::set_boxed_logger(Box::new(Logger::default())).unwrap());
+        INIT.call_once(|| {
+            // silently ignore errors here, as running tests will hit this condition for some reason.
+            log::set_boxed_logger(Box::new(Logger::default())).ok();
+        });
```
Comment (flub) [inline sentry-log/src/integration.rs:48]: So tests end up running this more than once?  I'm not convinced this is the right work around in this case

File: sentry/src/transport.rs [hunk 6]
```diff
@@ -222,9 +222,12 @@ implement_http_transport! {
                         }
                     }
 
+                    let mut body = vec![];
```
Comment (flub) [inline sentry/src/transport.rs:225]: Out of curiosity, I tend to write this as `Vec::new()` but you use the macro, are there any advantages to using the macro?  (my choice is purely because I like code with fewer macros)

File: sentry/src/transport.rs [hunk 7]
```diff
@@ -222,9 +222,12 @@ implement_http_transport! {
...
+                    envelope.to_writer(&mut body).unwrap();
```
Comment (flub) [inline sentry/src/transport.rs:226]: The `envelope.to_writer()` call also ends up calling some serde stuff that returns results instead of only writing into the vec (which is infallible).  Are you sure it's safe to unwrap those errors?

Comment (flub) [reply sentry/src/transport.rs:226]: (Yes, I am aware that differs from my comment on the test code of the envelope, but this isn't test code i gues)

File: sentry/src/transport.rs [hunk 8]
```diff
@@ -325,7 +328,10 @@ implement_http_transport! {
                     _ => {}
                 }
 
-                let mut body = Cursor::new(serde_json::to_vec(&event).unwrap());
+                let mut body = vec![];
+                envelope.to_writer(&mut body).unwrap();
```
Comment (flub) [inline sentry/src/transport.rs:332]: same about errors i think

File: sentry/src/transport.rs [hunk 9]
```diff
@@ -440,25 +446,16 @@ implement_http_transport! {
                         }
                     }
 
-                    let req_result = http_client
+                    let mut body = vec![];
+                    envelope.to_writer(&mut body).unwrap();
```
Comment (flub) [inline sentry/src/transport.rs:450]: same again i think

File: sentry-core/src/envelope.rs (hunk 2 above)
Comment (Swatinem) [reply sentry-core/src/envelope.rs:129]: The `event_id` is not nullable. I can add more tests once I add session support.

File: sentry-core/src/envelope.rs (hunk 4 above)
Comment (Swatinem) [reply sentry-core/src/envelope.rs:3]: you mean the `protocol` module? Not really sure, but we can move it there later on as well.

File: sentry/src/transport.rs (hunk 7 above)
Comment (Swatinem) [reply sentry/src/transport.rs:226]: Actually there is so much unwrapping going on in the transports, I don’t think its worth caring about this I think.

File: sentry-log/src/integration.rs (hunk 5 above)
Comment (Swatinem) [reply sentry-log/src/integration.rs:48]: So this was super weird. Some of my local testruns were crashing hard. The global logger is not thread safe, so I suspect thats what was causing things to fail, though not sure.

File: sentry-core/src/envelope.rs (hunk 4 above)
Comment (flub) [reply sentry-core/src/envelope.rs:3]: yeah, meant the protocol module.  I guess need to decide on this before releasing it though
